### ISSUE TITLE
[Feature]: Adding attention mask to vllm.attention.Attention

### ISSUE THREAD

**Issue by jinan-zhou** [vllm-project/vllm#17869]
### 🚀 The feature, motivation and pitch

Right now, `vllm.attention.Attention` implements the standard self-attention operation:
```math
 \text{Attention}(Q, K, V) = \text{softmax}\left( \frac{QK^\top}{\sqrt{d_k}} \right) V
```

However, it is very common among models to include an attention mask $$M$$:
```math
 \text{Attention}(Q, K, V, M) = \text{softmax}\left( \frac{QK^\top + M}{\sqrt{d_k}} \right) V
```

Like the `attn_mask` parameter of [`torch.nn.functional.scaled_dot_product_attention`](https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html).

I am working on adding a new model, but found the lack of attention mask to be the biggest blocker. Are there any work-arounds at this moment? If not, what is the best way to add attention mask to vLLM? Any comments are appreciated!

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by jinan-zhou** [comment]
For anyone with the same concern, the answer to the first question is negative. Let me paste the answer from the bot in vLLM Slack:
> Currently, vLLM does not fully support passing custom attention masks (such as arbitrary 4D attention masks) to its optimized attention backends like FlashAttention or FlashInfer. Only the XFormers backend allows some flexibility, but even there, custom attention bias (which can be used to implement custom masks) is only supported in the prefill phase, not in decode, and is not exposed as a general API for arbitrary masks. This limitation is discussed in detail in [RFC: Encoder/decoder models & feature compatibility](https://github.com/vllm-project/vllm/issues/7366) and [4D Attention Mask issue](https://github.com/vllm-project/vllm/issues/6615).
> 
> If you need to use a custom attention mask, you would have to modify the backend kernels or use a less optimized backend (like PyTorch SDPA) with significant performance trade-offs.

**Comment by wulaoshi** [comment]
Hello, I have the same requirement. Have you implemented it? How did you achieve it?
